airflow/models/dag.py

```diff
@@ -1378,6 +1379,33 @@ def normalized_schedule_interval(self) -> ScheduleInterval:
             _schedule_interval = self.schedule_interval
         return _schedule_interval
 
+    @staticmethod
+    @internal_api_call
+    @provide_session
+    def fetch_callback(
+        dag: DAG, dagrun: DagRun | DagRunPydantic, success=True, reason=None, session=NEW_SESSION
+    ):
+        """
+        Fetch the appropriate callbacks depending on the value of success, namely the
+        on_failure_callback or on_success_callback. This method gets the context of a
+        single TaskInstance part of this DagRun and returns it along the list of callbacks.
+
+        :param dag: DAG object
+        :param dagrun: DagRun object
+        :param success: Flag to specify if failure or success callback should be called
+        :param reason: Completion reason
+        :param session: Database session
+        """
+        callbacks = dag.on_success_callback if success else dag.on_failure_callback
+        if callbacks:
+            callbacks = callbacks if isinstance(callbacks, list) else [callbacks]
+            tis = dagrun.get_task_instances(session=session)
+            ti = tis[-1]  # get first TaskInstance of DagRun
```

Comment:
This is unnecessarily costly. Doing a `session...one()` here would be better.

Reply:
This is the current code. See [here](https://github.com/apache/airflow/blob/main/airflow/models/dag.py#L1401). Happy to optimize this one on a separate PR but I think this PR is complex enough to not address it here :)

Reply:
Agree. Let's not mix the twol